Add SweepGradient

xqwzts opened the pull request:
This PR adds a `SweepGradient` class, extending `Gradient` to expose the engine's `ui.Gradient.sweep` shader.

Similar to `LinearGradient` and `RadialGradient` - `SweepGradient`s can be used in a `BoxDecoration` or passed to a `Paint`'s `shader`.

xqwzts committed:
SweepGradient

xqwzts committed:
Use radians for startAngle and endAngle

xqwzts committed:
Improve documentation and add example

xqwzts committed:
Add SweepGradient tests

xqwzts committed:
Add self to authors

xqwzts committed:
Remove trailing semi-colon from sample code

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +655,232 @@ class RadialGradient extends Gradient {
+  }
+
+  /// Returns a new [SweepGradient] with its properties (in particular the
+  /// colors) scaled by the given factor.
```
Really, it's only the color that is scaled, not any other properties. It also does include scaling the alpha component too, which some people might not expect.  You might explain that in order to explain why a scale of zero yields a transparent color instead of black.

xqwzts replied:
Reworded to the following to make it clear that only the alpha component is scaled: (that's what happens when `Color.lerp(null, color)` is called):

  /// Returns a new [SweepGradient] with its colors scaled by the given factor.
  ///
  /// Since the alpha component of the Color is what is scaled, a factor
  /// of 0.0 or less results in a gradient that is fully transparent.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +655,232 @@ class RadialGradient extends Gradient {
+
+  /// Linearly interpolate between two [SweepGradient]s.
+  ///
+  /// If either gradient is null, this function linearly interpolates from a
```
I'd say this differently: 
If either gradient is null, then the non-null gradient is returned with its color scaled in the same way as the [scale] function.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +655,232 @@ class RadialGradient extends Gradient {
+  ///
+  /// If neither gradient is null, they must have the same number of [colors].
+  ///
+  /// The `t` argument represents position on the timeline, with 0.0 meaning
```
represents position --> represents a position

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +655,232 @@ class RadialGradient extends Gradient {
+      return false;
+    if (colors != null) {
+      assert(typedOther.colors != null);
+      assert(colors.length == typedOther.colors.length);
```
Shouldn't these both just return false instead of asserting?  If their colors are null or not the same length, then they're just not equal.

xqwzts replied:
I'm doing that here to stay consistent with the LinearGradient and RadialGradient implementations, I'm not sure why they assert on color and stops differences though. [That was introduced in this commit if it helps make sense of it: https://github.com/Hixie/flutter/commit/552f26d71324e559e69a38e5ecc16c3110c580ab#diff-1756a3cb685f68e4d8f389736fda0df2R379].

My best guess is it was added so that attempting to animated between gradients with different colors/stops would assert [although  `lerp` would assert in that case, so it seems redundant?]

gspencergoog replied:
For consistency, I guess just leave them there, then, and I'll put together another PR that will remove them all after talking it over with @Hixie as to why they're in there.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +655,232 @@ class RadialGradient extends Gradient {
+      }
+    }
+    if (stops != null) {
+      assert(typedOther.stops != null);
```
same here for the stops asserts: shouldn't they just return false?

gspencergoog commented on packages/flutter/test/painting/gradient_test.dart:
```diff
@@ -263,6 +303,82 @@ void main() {
     ));
   });
 
+  test('SweepGradient lerp test', () {
```
What about a scale test too?

xqwzts commented:
Thanks for the review @gspencergoog , I'm hoping to go through all your comments in the next few days :)

xqwzts committed:
Add scale test

xqwzts committed:
Merge branch 'master' into gradient_sweep

xqwzts committed:
Various dartdoc fixes

xqwzts committed:
Merge branch 'gradient_sweep' of github.com:xqwzts/flutter into gradient_sweep

xqwzts commented:
@gspencergoog Updated, PTAL when you can.

I've addressed your doc comments and added a scale test. I'm not sure about the colors/stops assertions in the `==` implementation though, could you let me know what you'd like me to do there? (More details in [this](https://github.com/flutter/flutter/pull/17368#discussion_r191254107) comment).

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +656,230 @@ class RadialGradient extends Gradient {
+
+/// A 2D sweep gradient.
+///
+/// This class is used by [BoxDecoration] to represent gradients. This abstracts
```
represent gradients -> represent sweep gradients

xqwzts replied:
Done.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +656,230 @@ class RadialGradient extends Gradient {
+  /// For example, an alignment of (0.0, 0.0) will place the sweep
+  /// gradient in the center of the box.
+  ///
+  /// If this is a [Alignment], then it is expressed as a vector from
```
a [Alignment] --> an [Alignment]

xqwzts replied:
Done.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +656,230 @@ class RadialGradient extends Gradient {
+  /// coordinate (0.0, 0.0), in a coordinate space that maps the center of the
+  /// paint box at (0.0, 0.0) and the bottom right at (1.0, 1.0).
+  ///
+  /// It can also be a [AlignmentDirectional], where the start is the left in
```
a [AlignmentDirectional] --> an [AlignmentDirectional]

xqwzts replied:
Done.

gspencergoog commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -651,3 +656,230 @@ class RadialGradient extends Gradient {
+  ///
+  /// For details, see [TileMode].
+  ///
+  /// ![](https://flutter.github.io/assets-for-api-docs/dart-ui/tile_mode_clamp_sweep.png)
```
These URLs need to have an "assets/" added (it changed recently):
https://flutter.github.io/assets-for-api-docs/assets/dart-ui/tile_mode_clamp_sweep.png
https://flutter.github.io/assets-for-api-docs/assets/dart-ui/tile_mode_mirror_sweep.png
https://flutter.github.io/assets-for-api-docs/assets/dart-ui/tile_mode_repeated_sweep.png

xqwzts replied:
Done.

gspencergoog commented:
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)
(modulo the doc comment changes I commented on)

xqwzts committed:
Various doc fixes

xqwzts commented:
Thanks for giving it another look @gspencergoog .

Updated the docs. PTAL when you can.

xqwzts commented:
Thanks for all the help @gspencergoog .
